servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpConnector.java

```diff
@@ -68,67 +76,74 @@ private TcpConnector() {
      * @param config The {@link ReadOnlyTcpClientConfig} to use while connecting.
      * @param autoRead if {@code true} auto read will be enabled for new {@link Channel}s.
      * @param executionContext The {@link ExecutionContext} to use for the returned {@link NettyConnection}.
+     * @param connectionFactory {@link Function} to create a {@link NettyConnection} asynchronously.
+     * @param <C> Type of the created connection.
      * @return A {@link Single} that completes with a new {@link Channel} when connected.
      */
-    public static Single<Channel> connect(@Nullable SocketAddress localAddress, Object resolvedRemoteAddress,
-                                          ReadOnlyTcpClientConfig config, boolean autoRead,
-                                          ExecutionContext executionContext) {
+    public static <C extends ListenableAsyncCloseable> Single<C> connect(
+            final @Nullable SocketAddress localAddress, final Object resolvedRemoteAddress,
+            final ReadOnlyTcpClientConfig config, final boolean autoRead, final ExecutionContext executionContext,
+            final Function<Channel, Single<? extends C>> connectionFactory) {
         requireNonNull(resolvedRemoteAddress);
         requireNonNull(config);
         requireNonNull(executionContext);
-        return new SubscribableSingle<Channel>() {
+        return new SubscribableSingle<C>() {
             @Override
-            protected void handleSubscribe(final Subscriber<? super Channel> subscriber) {
-                connectFutureToListener(localAddress, resolvedRemoteAddress, subscriber,
-                        connect0(localAddress, resolvedRemoteAddress, config, autoRead, executionContext, subscriber));
+            protected void handleSubscribe(final Subscriber<? super C> subscriber) {
+                ConnectHandler<C> connectHandler = new ConnectHandler<>(subscriber, connectionFactory);
+                try {
+                    Future<?> connectFuture = connect0(localAddress, resolvedRemoteAddress, config, autoRead,
+                            executionContext, connectHandler);
+                    connectHandler.connectFuture(connectFuture);
+                    connectFuture.addListener(f -> {
```

Comment:
Consider verifying if `connectFuture` is already done before applying the listener, as we do in other places:
https://github.com/apple/servicetalk/blob/e5f4c563e0b5818fc8415d64a3ddd1ba1b90b393/servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscoverer.java#L370-L376

Reply:
We do that when we are already on the eventloop and no ordering guarantees are required. In this case, adding a listener makes sure that the `Single` is completed on the eventloop.

(This code was just refactored, so I would continue with the approach what was present before but if you feel we need to change this in general in all places, we can discuss/do that separately)